jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/CliDownloaderUtil.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.common.util;
+
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.http.HttpClient;
+import io.fabric8.kubernetes.client.http.HttpResponse;
+import io.fabric8.kubernetes.client.utils.HttpClientUtils;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.commons.io.FilenameUtils.removeExtension;
+import static org.eclipse.jkube.kit.common.archive.ArchiveDecompressor.extractArchive;
+
+public class CliDownloaderUtil {
+  private CliDownloaderUtil() { }
+
+  public static String downloadCli(String baseDownloadUrl, String cliBinaryPrefix, String cliArtifactName, File outputDirectory) throws IOException {
+    URL downloadUrl = new URL(String.format("%s/%s", baseDownloadUrl, cliArtifactName));
+    File targetExtractionDir = outputDirectory.toPath().resolve(removeExtension(cliArtifactName)).toFile();
+
+    downloadAndExtractTo(downloadUrl, targetExtractionDir);
+
+    return getCliBinaryPathFromExtractedDir(targetExtractionDir, cliBinaryPrefix);
+  }
+
+  private static void downloadAndExtractTo(URL downloadUrl, File target) throws IOException {
+    try (HttpClient client = HttpClientUtils.getHttpClientFactory().newBuilder(Config.empty()).build()) {
+      final HttpResponse<InputStream> response = client.sendAsync(
+              client.newHttpRequestBuilder().timeout(30, TimeUnit.MINUTES).url(downloadUrl).build(), InputStream.class)
+          .get();
+      if (!response.isSuccessful()) {
+        throw new IOException("Server returned (" + response.code() + ") while downloading " + downloadUrl);
+      }
+      try (InputStream is = response.body()) {
+        extractArchive(is, target);
+      }
+    } catch(InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IOException("Download interrupted", ex);
+    } catch (IOException | ExecutionException e) {
+      throw new IOException("Failed to download URL " + downloadUrl + " to  " + target + ": " + e, e);
+    }
+  }
+
+  private static String getCliBinaryPathFromExtractedDir(File targetExtractionDir, String binaryPrefix) {
+    String cliPath = null;
+    File[] cliArtifactList = targetExtractionDir.listFiles(f -> f.getName().startsWith(binaryPrefix));
+    if (cliArtifactList != null && cliArtifactList.length >= 1) {
+      cliPath = cliArtifactList[0].getAbsolutePath();
+    }
+    return cliPath;
+  }
```

Comment:
Is this behavior consistent for any CLI we decide to download using this util class?

Reply:
I was trying with `canExecute` fileter but it doesn't seem to work. We seem to be losing permissions while decompressing archive .

Reply:
There is also option to select file with largest size.